src/DotVVM.Framework/Compilation/Parser/Binding/Parser/BindingParser.cs

```diff
@@ -340,7 +340,149 @@ private BindingParserNode ReadUnaryExpression()
+            Read();
+            SkipWhiteSpace();
+            ClearRestorePoint();
+
+            var body = ReadLambdaBodyExpression();
+            return CreateNode(new LambdaBindingParserNode(parameters, body), startIndex);
+        }
+
+        private bool TryReadLambdaParametersExpression(out List<LambdaParameterBindingParserNode> parameters)
+        {
+            var startIndex = CurrentIndex;
+            var waitingForParameter = false;
+            parameters = new List<LambdaParameterBindingParserNode>();
+            if (Peek()?.Type == BindingTokenType.OpenParenthesis)
+            {
+                // Begin parameters parsing - read opening parenthesis
+                Read();
+                SkipWhiteSpace();
+
+                while (Peek()?.Type != BindingTokenType.CloseParenthesis)
+                {
+                    // Try read parameter definition (either implicitly defined type or explicitely)
+                    if (!TryReadLambdaParameterDefinition(out var typeDef, out var nameDef))
+                        return false;
+                    parameters.Add(new LambdaParameterBindingParserNode(typeDef, nameDef!));
+                    waitingForParameter = false;
+
+                    if (Peek()?.Type == BindingTokenType.Comma)
+                    {
+                        Read();
+                        SkipWhiteSpace();
+                        waitingForParameter = true;
+                    }
+                    else
+                    {
+                        // If next is not comma then we must be finished
+                        break;
+                    }
+                }
+
+                // End parameters parsing - read closing parenthesis
+                if (Peek()?.Type != BindingTokenType.CloseParenthesis)
+                    return false;
+                Read();
+                SkipWhiteSpace();
+            }
+            else
+            {
+                // Support lambdas with single implicit parameter and no parentheses: arg => Method(arg)
+                var parameter = ReadIdentifierExpression(false);
+                if (parameter.HasNodeErrors)
+                    return false;
+
+                parameters.Add(new LambdaParameterBindingParserNode(null, CreateNode(parameter, startIndex)));
+            }
+
+            if (waitingForParameter)
+                return false;
+
+            return true;
+        }
+
+        private bool TryReadLambdaParameterDefinition(out BindingParserNode? type, out BindingParserNode? name)
+        {
+            name = null;
+            type = null;
+            if (Peek()?.Type != BindingTokenType.Identifier)
+                return false;
+
+            type = ReadIdentifierExpression(true);
+            if (Peek()?.Type != BindingTokenType.Identifier)
+            {
+                name = type;
+                type = null;
+                return true;
+            }
+            name = ReadIdentifierExpression(true);
```

Comment:
We should check that name is a SimpleNameBindingParserNode - it should not have generic arguments and so on.